File: pkg/level/generator/object/TreeParityTest.java
// TreeParityTest.java - 验证Go实现与Java 1.12树木生成的数学一致性

import java.util.Random;

/**
 * TreeParityTest - 验证Go实现与Java 1.12树木生成的数学一致性
 * 
 * 测试项目:
 * 1. 深色橡木高度公式: nextInt(3) + nextInt(2) + 6
 * 2. 橡木角落逻辑: Math.abs(l1) != j1 || Math.abs(j2) != j1 || (rand.nextInt(2) != 0
 * && i4 != 0)
 * 3. 半径计算: 1 - i4/2 (整数除法)
 */
public class TreeParityTest {

    public static void main(String[] args) {
        long seed = 12345L;
        Random rand = new Random(seed);

        System.out.println("=== Java 1.12 Tree Generation Parity Test ===");
        System.out.println("Seed: " + seed);
        System.out.println();

        // Test 1: Dark Oak Height Formula
        System.out.println("--- Test 1: Dark Oak Height Formula ---");
        System.out.println("Formula: nextInt(3) + nextInt(2) + 6");
        rand = new Random(seed);
        for (int i = 0; i < 20; i++) {
            int height = rand.nextInt(3) + rand.nextInt(2) + 6;
            System.out.println("Tree " + i + ": height=" + height);
        }
        System.out.println();

        // Test 2: Oak Leaf Radius Calculation (Integer Division)
        System.out.println("--- Test 2: Leaf Radius (Integer Division) ---");
        System.out.println("Formula: j1 = 1 - i4/2");
        for (int i4 = -3; i4 <= 0; i4++) {
            int j1 = 1 - i4 / 2;
            System.out.println("i4=" + i4 + " -> j1=" + j1);
        }
        System.out.println();

        // Test 3: Corner Rounding Logic
        System.out.println("--- Test 3: Corner Rounding Logic ---");
        System.out.println("Formula: place if (abs(l1)!=j1 || abs(j2)!=j1 || (nextInt(2)!=0 && i4!=0))");
        rand = new Random(seed);
        int treeHeight = 5;
        int placedCount = 0;
        int skippedCount = 0;

        for (int i3 = 0 - 3 + treeHeight; i3 <= 0 + treeHeight; i3++) {
            int i4 = i3 - (0 + treeHeight);
            int j1 = 1 - i4 / 2;

            for (int k1 = 0 - j1; k1 <= 0 + j1; k1++) {
                int l1 = k1 - 0;

                for (int i2 = 0 - j1; i2 <= 0 + j1; i2++) {
                    int j2 = i2 - 0;

                    // The actual Java 1.12 condition
                    boolean shouldPlace = Math.abs(l1) != j1 || Math.abs(j2) != j1 ||
                            (rand.nextInt(2) != 0 && i4 != 0);

                    if (shouldPlace) {
                        placedCount++;
                    } else {
                        skippedCount++;
                        System.out.println("SKIP: i4=" + i4 + " j1=" + j1 + " l1=" + l1 + " j2=" + j2);
                    }
                }
            }
        }
        System.out.println("Total placed: " + placedCount + ", skipped: " + skippedCount);
        System.out.println();

        // Test 4: RNG Sequence for verification
        System.out.println("--- Test 4: RNG Sequence (first 50 values) ---");
        rand = new Random(seed);
        System.out.print("nextInt(100): ");
        for (int i = 0; i < 10; i++) {
            System.out.print(rand.nextInt(100) + " ");
        }
        System.out.println();

        rand = new Random(seed);
        System.out.print("nextBool: ");
        for (int i = 0; i < 10; i++) {
            System.out.print(rand.nextBoolean() + " ");
        }
        System.out.println();

        System.out.println();
        System.out.println("=== Test Complete ===");
        System.out.println("Run the corresponding Go test and compare outputs.");
    }
}

File: pkg/level/generator/object/tree_parity_test.go
package object

import (
	"fmt"
	"math"
	"testing"

	"github.com/scaxe/scaxe-go/pkg/math/rand"
)

func TestTreeParity(t *testing.T) {
	seed := int64(12345)

	fmt.Println("=== Go Tree Generation Parity Test ===")
	fmt.Println("Seed:", seed)
	fmt.Println()

	fmt.Println("--- Test 1: Dark Oak Height Formula ---")
	fmt.Println("Formula: NextBoundedInt(3) + NextBoundedInt(2) + 6")
	r := rand.NewRandom(seed)
	for i := 0; i < 20; i++ {
		height := r.NextBoundedInt(3) + r.NextBoundedInt(2) + 6
		fmt.Printf("Tree %d: height=%d\n", i, height)
	}
	fmt.Println()

	fmt.Println("--- Test 2: Leaf Radius (Integer Division) ---")
	fmt.Println("Formula: j1 = 1 - i4/2")
	for i4 := -3; i4 <= 0; i4++ {
		j1 := 1 - i4/2
		fmt.Printf("i4=%d -> j1=%d\n", i4, j1)
	}
	fmt.Println()

	fmt.Println("--- Test 3: Corner Rounding Logic ---")
	fmt.Println("Formula: place if (abs(l1)!=j1 || abs(j2)!=j1 || (NextBoundedInt(2)!=0 && i4!=0))")
	r = rand.NewRandom(seed)
	treeHeight := 5
	placedCount := 0
	skippedCount := 0

	for i3 := 0 - 3 + treeHeight; i3 <= 0+treeHeight; i3++ {
		i4 := i3 - (0 + treeHeight)
		j1 := 1 - i4/2

		for k1 := 0 - j1; k1 <= 0+j1; k1++ {
			l1 := k1 - 0

			for i2 := 0 - j1; i2 <= 0+j1; i2++ {
				j2 := i2 - 0

				absL1 := int(math.Abs(float64(l1)))
				absJ2 := int(math.Abs(float64(j2)))

				shouldPlace := absL1 != j1 || absJ2 != j1 ||
					(r.NextBoundedInt(2) != 0 && i4 != 0)

				if shouldPlace {
					placedCount++
				} else {
					skippedCount++
					fmt.Printf("SKIP: i4=%d j1=%d l1=%d j2=%d\n", i4, j1, l1, j2)
				}
			}
		}
	}
	fmt.Printf("Total placed: %d, skipped: %d\n", placedCount, skippedCount)
	fmt.Println()

	fmt.Println("--- Test 4: RNG Sequence (first 50 values) ---")
	r = rand.NewRandom(seed)
	fmt.Print("NextBoundedInt(100): ")
	for i := 0; i < 10; i++ {
		fmt.Print(r.NextBoundedInt(100), " ")
	}
	fmt.Println()

	r = rand.NewRandom(seed)
	fmt.Print("NextBoolean: ")
	for i := 0; i < 10; i++ {
		fmt.Print(r.NextBoolean(), " ")
	}
	fmt.Println()

	fmt.Println()
	fmt.Println("=== Test Complete ===")
	fmt.Println("Compare with Java TreeParityTest output.")

	t.Run("DarkOakHeightRange", func(t *testing.T) {
		r := rand.NewRandom(seed)
		for i := 0; i < 1000; i++ {
			height := r.NextBoundedInt(3) + r.NextBoundedInt(2) + 6
			if height < 6 || height > 10 {
				t.Errorf("Dark Oak height out of range: %d (expected 6-10)", height)
			}
		}
	})

	t.Run("IntegerDivision", func(t *testing.T) {

		expected := map[int]int{-3: 2, -2: 2, -1: 1, 0: 1}
		for i4, expectedJ1 := range expected {
			j1 := 1 - i4/2
			if j1 != expectedJ1 {
				t.Errorf("i4=%d: got j1=%d, want %d", i4, j1, expectedJ1)
			}
		}
	})
}

func TestDarkOakHeightDistribution(t *testing.T) {
	counts := make(map[int]int)
	r := rand.NewRandom(42)

	for i := 0; i < 10000; i++ {
		height := r.NextBoundedInt(3) + r.NextBoundedInt(2) + 6
		counts[height]++
	}

	fmt.Println("Dark Oak Height Distribution (10000 samples):")
	for h := 6; h <= 10; h++ {
		pct := float64(counts[h]) / 100.0
		fmt.Printf("  Height %d: %d (%.1f%%)\n", h, counts[h], pct)
	}

	if counts[7] <= counts[6] {
		t.Logf("Warning: Height 7 should be more common than 6 (got %d vs %d)", counts[7], counts[6])
	}
}
